colexechash: minor improvements and cleanup

yuzefovich opened the pull request:
**execgen: skip encoding/decoding JSON when hashing it**

Previously, in order to hash a JSON object we would encode and decode it
(due to the behavior of `coldata.JSONs.Get` and the setup of the
execgen). This commit refactors how the hash functions are handled in
the execgen which allows us to peek inside of `coldata.Bytes` (that is
under the `coldata.JSONs`) in order to get direct access to the
underlying `[]byte`. This should be a minor performance improvement but
also allows us to remove the need for the overload helper in the hashing
context.

It is possible (although I'm not certain) that the hash computation is
now different, so this commit bumps the DistSQL versions to be safe.

Release note: None

**colexechash: fix BCEs in some cases**

Previously, we were not getting (and not asserting) some of the bounds
check eliminations in `rehash` function because we were accessing
`.Sliceable` property in the wrong context. This is now fixed (by
accessing it via `.Global`) as well as assertions are added correctly.
Additionally, this commit pulls out the call to the cancel checker out
of the templated block to reduce the amount of code duplication.

Release note: None

yuzefovich committed:
execgen: skip encoding/decoding JSON when hashing it

Previously, in order to hash a JSON object we would encode and decode it
(due to the behavior of `coldata.JSONs.Get` and the setup of the
execgen). This commit refactors how the hash functions are handled in
the execgen which allows us to peek inside of `coldata.Bytes` (that is
under the `coldata.JSONs`) in order to get direct access to the
underlying `[]byte`. This should be a minor performance improvement but
also allows us to remove the need for the overload helper in the hashing
context.

It is possible (although I'm not certain) that the hash computation is
now different, so this commit bumps the DistSQL versions to be safe.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74702)
<!-- Reviewable:end -->

yuzefovich committed:
colexechash: fix BCEs in some cases

Previously, we were not getting (and not asserting) some of the bounds
check eliminations in `rehash` function because we were accessing
`.Sliceable` property in the wrong context. This is now fixed (by
accessing it via `.Global`) as well as assertions are added correctly.
Additionally, this commit pulls out the call to the cancel checker out
of the templated block to reduce the amount of code duplication.

Release note: None

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 11 of 11 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74702#-:-MtEJV9Q4Bh4IAa_ADUe:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads_hash.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74702#-MtEH_PT8dcDI4bJIjeQ:-MtEH_PUEKtvOvAhRfJ_:b-fmiu0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c5ad265122b1fbf3f89a284aa7e6eedd7c6a3c5/pkg/sql/colexec/execgen/cmd/execgen/overloads_hash.go#L150)):*
> ```Go
>         // Access the underlying []byte directly which allows us to skip
>         // decoding-encoding of the JSON object.
>         _b := %[2]s.Bytes.Get(%[3]s)
> ```

What kind of notation is this (i.e. "%[2]s", "%[3]s")?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74702#-:-MtELirYBxJ0ltIH01bS:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rharding6373)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads_hash.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74702#-MtEH_PT8dcDI4bJIjeQ:-MtELDcT5_T1J7E4BGeD:b-ctnhqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c5ad265122b1fbf3f89a284aa7e6eedd7c6a3c5/pkg/sql/colexec/execgen/cmd/execgen/overloads_hash.go#L150)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

What kind of notation is this (i.e. "%[2]s", "%[3]s")?
</blockquote></details>

`%[1]s` means "use the first argument after the format string and print it as `string` type", `%[2]s` ditto for the second argument after for the format string, etc. The idea is that this notation allows you to reuse the same argument many times and change order if necessary (although in this particular instance we could have avoided using this notation).


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74702)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads_hash.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74702#-MtEH_PT8dcDI4bJIjeQ:-MtESeK80B7Ex3KzbyMD:bj446d) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c5ad265122b1fbf3f89a284aa7e6eedd7c6a3c5/pkg/sql/colexec/execgen/cmd/execgen/overloads_hash.go#L150)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`%[1]s` means "use the first argument after the format string and print it as `string` type", `%[2]s` ditto for the second argument after for the format string, etc. The idea is that this notation allows you to reuse the same argument many times and change order if necessary (although in this particular instance we could have avoided using this notation).
</blockquote></details>

Ah, for some reason I didn't realize that this was in a `fmt.Sprintf`. Makes sense now, thanks.


<!-- Sent from Reviewable.io -->
